src/providers/completer/reference.ts

```diff
@@ -1,114 +1,109 @@
+        // Extract cached references
+        const refList: string[] = []
+        Object.keys(this.extension.manager.cachedContent).forEach(cachedFile => {
+            const cachedRefs = this.extension.manager.cachedContent[cachedFile].element.reference
+            if (cachedRefs === undefined) {
+                return
+            }
+            for (const ref of cachedRefs) {
+                if (ref.range === undefined) {
+                    continue
+                }
+                this.suggestions[ref.label] = {...ref,
+                    file: cachedFile,
+                    position: ref.range.start,
+                    range: args.document.getWordRangeAtPosition(args.position, /[-a-zA-Z0-9_:.]+/),
+                }
+                refList.push(ref.label)
             }
         })
-        Object.keys(references).forEach((key) => {
-            this.referenceData[key] = {
-                item: references[key],
-                text: references[key].text,
-                file: filePath
+        // Remove references that has been deleted
+        Object.keys(this.suggestions).forEach(key => {
+            if (refList.indexOf(key) <= -1) {
+                delete this.suggestions[key]
             }
         })
+        // Compile the suggestion object to array
+        return Object.keys(this.suggestions).map(key => this.suggestions[key])
     }
 
-    getReferenceItems(content: string) {
-        const itemReg = /(?:\\label(?:\[[^[\]{}]*\])?|(?:^|[,\s])label=){([^}]*)}/gm
-        const items: {[key: string]: ReferenceEntry['item']} = {}
-        content = stripComments(content, '%')
-        const noELContent = content.split('\n').filter(para => para !== '').join('\n')
-        while (true) {
-            const result = itemReg.exec(content)
-            if (result === null) {
-                break
-            }
-            if (!(result[1] in items)) {
-                const prevContent = noELContent.substring(0, noELContent.substring(0, result.index).lastIndexOf('\n') - 1)
-                const followLength = noELContent.substring(result.index, noELContent.length).split('\n', 4).join('\n').length
-                const positionContent = content.substring(0, result.index).split('\n')
-                items[result[1]] = {
-                    reference: result[1],
-                    text: `${noELContent.substring(prevContent.lastIndexOf('\n') + 1, result.index + followLength)}\n...`,
-                    position: new vscode.Position(positionContent.length - 1, positionContent[positionContent.length - 1].length)
+    update(file: string, content: string) {
+        const refs = this.getReferenceFromAST(latexParser.parse(content), content.split('\n'))
+        this.extension.manager.cachedContent[file].element.reference = refs
+    }
+
+    private getReferenceFromAST(ast: latexParser.LatexAst | latexParser.Node, lines: string[], nextNode?: latexParser.Node): vscode.CompletionItem[] {
```

Comment:
`LatexAst` and `Node` are completely different. Please split `getReferenceFromAST` defining `getReferenceFromNode`.

Reply:
There are two classes: command/string, and others (node and root). So which one is AST?

Reply:
How about the following? I have not tested it yet.  [branch](https://github.com/tamuratak/LaTeX-Workshop/tree/completer-rework_2).

```typescript
    update(file: string, content: string) {
        const refs = this.getReferenceFromNodeArray(latexParser.parse(content).content, content.split('\n'))
        this.extension.manager.cachedContent[file].element.reference = refs
    }

    private getReferenceFromNodeArray(nodes: latexParser.Node[], lines: string[]): vscode.CompletionItem[] {
        let refs: vscode.CompletionItem[] = []
        for (let index = 0; index < nodes.length; ++index) {
            if (index < nodes.length - 1) {
                // Also pass the next node to handle cases like `label={some-text}`
                refs = refs.concat(this.getReferenceFromNode(nodes[index], lines, nodes[index+1]))
            } else {
                refs = refs.concat(this.getReferenceFromNode(nodes[index], lines))
            }
        }
        return refs
    }

    private getReferenceFromNode(node: latexParser.Node, lines: string[], nextNode?: latexParser.Node): vscode.CompletionItem[] {
        let refs: vscode.CompletionItem[] = []
        let label = ''
        if (latexParser.isCommand(node) && node.name === 'label') {
            // \label{some-text}
            const group = node.args.filter(latexParser.isGroup)[0]
            label = group ? latexParser.stringify(group.content) : label
        } else if (latexParser.isTextString(node) && node.content === 'label=' && nextNode !== undefined) {
            // label={some=text}
            if (latexParser.isGroup(nextNode)) {
                label = latexParser.stringify(nextNode.content)
            }
        }
        if (label !== '' && (latexParser.isCommand(node) || latexParser.isTextString(node))) {
            refs.push({
                label,
                kind: vscode.CompletionItemKind.Reference,
                documentation: lines.slice(node.location.start.line - 2, node.location.end.line + 4).join('\n'),
                range: new vscode.Range(node.location.start.line - 1, node.location.start.column,
                                        node.location.end.line - 1, node.location.end.column)
            })
            // Here we abuse the definition of range to store the location of the reference definition
            return refs
        }
        if (latexParser.hasContentArray(node)) {
            refs = refs.concat(this.getReferenceFromNodeArray(node.content, lines))
        }
        return refs
    }
```